ruby/basics/hashes.md

```diff
@@ -1,51 +1,185 @@
+
+The other thing you might notice is that hashes are made up of two parts: *keys* and *values*. A *key* is more or less analogous to an array's *index*, in that it's the label on the outside of the bin that helps us find what we're looking for. The *value* is what gets stored at a particular *key*.
+
+So in our bins from up above, `"random_word"`, `"Dorothy's math test score"`, and `"Guess what, literally another bin"` are all *keys*. And because they behave the same as indexes in arrays, you can call their values the same way:
+
+```ruby
+hash["random_word"]
+  => "ahoy"
+```
+
+Notice that when we call the hash's value by key, the key goes inside a pair of *brackets*, like when you're calling an array by index.
+
+Of course, hashes don't only take strings as keys and values. Ruby is a pretty flexible language so you can jam any old thing in there and it'll work just fine:
+
+```ruby
+hash = { 9 => "nine", :six => 6 }
+hash[9]
+  => "nine"
+hash[:six]
+  => 6
+```
+
+## Creating Hashes
+
+You've already seen a hash being created *literally* (e.g. `hash = { :whatever => "???" }`), but you can also call good old `#new` on the `Hash` class. If you give the `#new` method an argument, it'll even set a default value for keys that you don't specify:
+
+```ruby
+hash = Hash.new
+hash["me"]
+  => nil
+hash = Hash.new("you")
+hash["me"]
+  => "you"
+hash["him"]
+  => "you"
+```
+
+## Manipulating data in a hash
+
+### Adding data
+
+You can add to a hash the same way that you can add to an array. The easiest way is to call the key and set the value like you would with any other variable. Think of it like finding the right bin by its label and dropping something into the bin.
+
+```ruby
+shoes = Hash.new
+shoes["summer"] = "sandals"
+shoes["winter"] = "boots"
+```
+
+### Accessing data
+
+You can get at that data the same way, e.g. `shoes["summer"] #=> "sandals"`. Hashes also come with a couple of sporty methods for finding what you need:
+
+```ruby
+shoes["summer"]
+  => "sandals"
+shoes.key("summer")
+  => "sandals"
+```
+
+### Removing data
+
+The simple way to delete data from a hash is simply to call the value by its key and set it to zero:
+
+```ruby
+shoes["winter"] = nil
+# And now we have no footwear in the snow.
+```
+
+The fun way to delete data from a hash is to use the hash's `#delete` method, which provides the cool functionality of returning the value from the key-value pair was deleted from the hash. So:
+
+```ruby
+shoes.delete("summer")
+  => "sandals"
+shoes
+  => {} # Empty hash, now that we have gotten rid of our shoes.
+```
+
+
+## Methods
+
+Hashes respond to just about all the same methods as arrays do, since they're both part of Ruby's *Enumerable* class. Since hashes have a few more moving parts, though, they behave a little differently. For example, whereas `array.each { |element| puts element }` will iterate through an array and print out each element in the array, hashes take two variables in the block, like so:
```

Comment:
The next lesson is going to go into basic enumerables for hashes and arrays, this could be moved over to that lesson to cut down the duplication between the lessons.